File: transform/vision/src/main/scala/com/intel/analytics/zoo/transform/vision/image/Transformer.scala
/*
 * Copyright 2016 The BigDL Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package com.intel.analytics.zoo.transform.vision.image

import com.intel.analytics.bigdl.dataset.{ChainedTransformer, Transformer}
import com.intel.analytics.bigdl.utils.RandomGenerator._
import com.intel.analytics.zoo.transform.vision.image.opencv.OpenCVMat

import scala.collection.Iterator

abstract class FeatureTransformer() extends Transformer[ImageFeature, ImageFeature] {

  import FeatureTransformer.logger

  private var outKey: Option[String] = None

  def setOutKey(key: String): this.type = {
    outKey = Some(key)
    this
  }

  protected def transformMat(feature: ImageFeature): Unit = {}

  def transform(feature: ImageFeature): ImageFeature = {
    if (!feature.isValid) return feature
    try {
      transformMat(feature)
      if (outKey.isDefined) {
        require(outKey.get != ImageFeature.mat, s"the output key should not equal to" +
          s" ${ImageFeature.mat}, please give another name")
        if (feature.contains(outKey.get)) {
          val mat = feature(outKey.get).asInstanceOf[OpenCVMat]
          feature.opencvMat().copyTo(mat)
        } else {
          feature(outKey.get) = feature.opencvMat().clone()
        }
      }
    } catch {
      case e: Exception =>
        val uri = if (feature.contains(ImageFeature.uri)) feature(ImageFeature.uri) else ""
        logger.warn(s"failed ${uri} in transformer ${getClass}")
        e.printStackTrace()
        feature.isValid = false
    }
    feature
  }

  override def apply(prev: Iterator[ImageFeature]): Iterator[ImageFeature] = {
    prev.map(transform)
  }

  // scalastyle:off methodName
  // scalastyle:off noSpaceBeforeLeftBracket
  def ->(other: FeatureTransformer): FeatureTransformer = {
    new ChainedFeatureTransformer(this, other)
  }

  // scalastyle:off methodName
  // scalastyle:off noSpaceBeforeLeftBracket
  override def ->[C](other: Transformer[ImageFeature, C]): Transformer[ImageFeature, C] = {
    new ChainedTransformer(this, other)
  }
}

object FeatureTransformer {
  val logger = Logger.getLogger(getClass)
}

class ChainedFeatureTransformer(first: FeatureTransformer, last: FeatureTransformer) extends
  FeatureTransformer {

  override def transform(prev: ImageFeature): ImageFeature = {
    last.transform(first.transform(prev))
  }
}


class RandomTransformer(transformer: FeatureTransformer, maxProb: Double)
  extends FeatureTransformer {
  override def transform(prev: ImageFeature): ImageFeature = {
    if (RNG.uniform(0, 1) < maxProb) {
      transformer.transform(prev)
    }
    prev
  }

  override def toString: String = {
    s"Random[${transformer.getClass.getCanonicalName}, $maxProb]"
  }
}

object RandomTransformer {
  def apply(transformer: FeatureTransformer, maxProb: Double): RandomTransformer =
    new RandomTransformer(transformer, maxProb)
}

File: pyzoo/zoo/automl/feature/transformer.py
#
# Copyright 2018 Analytics Zoo Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from abc import ABC, abstractmethod


class BaseTransformer(ABC):
    """
        Abstract Base class for basic feature transformers.
    """

    @abstractmethod
    def transform(self, inputs, is_train=False):
        """
        fit data with the input
        :param inputs: input to be fitted
        :param is_train: indicate whether in training mode
        :return:
        """

    def inverse_transform(self, target):
        """
        inverse transform target value into origin magnitude
        :param target: target to be inverse transformed
        :return:
        """
        return target

    def save(self, file_path):
        """
        save the status of Transformer
        :param file_path
        :return:
        """
        pass

    def restore(self, file_path):
        """
        restore the status saved
        :param file_path
        :return:
        """
        pass


class PreRollTransformer(BaseTransformer):
    """
    base class for pre_roll transformers.
    """

    def transform(self, inputs, transform_cols=None, is_train=False):
        """
        fit data with the input
        :param inputs: input data frame
        :param transform_cols: columns to be transformed.
        :param is_train: indicate whether in training mode
        :return:
        """
        raise NotImplementedError()

    def inverse_transform(self, target):
        """
        inverse transform target value into origin magnitude
        :param target: target to be inverse transformed
        :return:
        """
        return target

    def save(self, file_path):
        """
        save the status of Transformer
        :param file_path
        :return:
        """
        pass

    def restore(self, file_path):
        """
        restore the status saved
        :param file_path
        :return:
        """
        pass


class RollTransformer(BaseTransformer):
    """
    base class for roll transformers.
    """
    def __init__(self, horizon=1):
        """
        :param horizon: an int or a range or a list
        """
        self.horizon = horizon

    def transform(self, inputs, past_seq_len=2, is_train=False):
        """
        fit data with the input
        :param inputs: numpy array
        :param past_seq_len: the look back sequence length that need to unrolled
        :param is_train: indicate whether in training mode
        :return:
        """
        raise NotImplementedError()


class PostRollTransformer(BaseTransformer):
    """
    base class for post_roll transformers.
    """

    def transform(self, inputs, is_train=False):
        """
        fit data with the input
        :param inputs: (x, y)
        :param is_train: indicate whether in training mode
        :return: (x,y)
        """
        raise NotImplementedError()

    def inverse_transform(self, target):
        """
        inverse transform target value into origin magnitude
        :param target: target to be inverse transformed
        :return:
        """
        return target

    def save(self, file_path):
        """
        save the status of Transformer
        :param file_path
        :return:
        """
        pass

    def restore(self, file_path):
        """
        restore the status saved
        :param file_path
        :return:
        """
        pass
